cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -246,6 +289,41 @@ public void run() {
     }
   }
 
+  private void runInTransaction(WorkflowAction action, TransactionContext transactionContext) {
```

Comment:
Also there is a code duplication to make the datasets transaction aware (before initialize and run). we can make it as a part of common method. Also the process of making datasets transaction aware need to happen for destroy as well.